builder/builder.go

```diff
@@ -412,97 +427,454 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
+		} else {
+			buff.WriteString(treeBar)
+		}
+	}
+
+	// Add a mid bar or an end bar
+	if levelEnded[level-1] {
+		buff.WriteString(treeEnd)
+	} else {
+		buff.WriteString(treeMid)
+	}
+
+	// Add the actual value
+	buff.WriteString(value)
+
+	return buff.String()
+}
+
+func (b *Builder) bundleTreePrint(set bundleSet, bundle string, level int, levelEnded []bool) {
+	fmt.Println(b.buildTreePrintValue(set[bundle], level, levelEnded))
+
+	levelEnded = append(levelEnded, false)
+	last := len(set[bundle].DirectIncludes) - 1
+	for i, include := range set[bundle].DirectIncludes {
+		levelEnded[level] = i == last
+		b.bundleTreePrint(set, include, level+1, levelEnded)
+	}
+}
+
+type listType int
+
+// Enum of available list types
+const (
+	MixList      listType = iota // List bundles in the mix (with includes)
+	LocalList                    // List bundles available locally
+	UpstreamList                 // List bundles available upstream
+)
+
+// ListBundles prints out a bundle list in either a flat list or tree view
+func (b *Builder) ListBundles(listType listType, tree bool) error {
+	var bundles bundleSet
+	var upstreamBundles bundleSet
+	var err error
+
+	// Get the bundle sets used for processing
+	mixBundles, err := b.getMixBundlesListAsSet()
+	if err != nil {
+		return err
+	}
+	localBundles, err := b.getDirBundlesListAsSet(b.LocalBundleDir)
+	if err != nil {
+		return err
+	}
+
+	// Assign "top level" bundles
+	switch listType {
+	case MixList:
+		bundles = mixBundles
+	case LocalList:
+		bundles = localBundles
+	case UpstreamList:
+		// Lazy-loaded because it's only needed for UpstreamList's
+		upstreamBundles, err = b.getDirBundlesListAsSet(getUpstreamBundlesPath(b.UpstreamVer))
+		if err != nil {
+			return err
+		}
+		bundles = upstreamBundles
+	}
+
+	// Used for pretty-printing lists
+	var max int
+	for _, bundle := range bundles {
+		if len(bundle.Name) > max {
+			max = len(bundle.Name)
+		}
+	}
+
+	// Create the full, parsed set
+	set, err := b.getFullBundleSet(bundles)
+	if err != nil {
+		return err
+	}
+
+	if tree {
+		// Print the tree view
+		sorted := getBundleSetKeysSorted(bundles)
+		for _, bundle := range sorted {
+			b.bundleTreePrint(set, bundle, 0, make([]bool, 0))
+		}
+
+		return nil
+	}
+
+	// Print a flat list and return
+	switch listType {
+	case MixList:
+		// Print the full, parsed set
+		sorted := getBundleSetKeysSorted(set)
+		for _, bundle := range sorted {
+			val := fmt.Sprintf("%-*s", max, set[bundle].Name)
+
+			if _, exists := localBundles[set[bundle].Name]; exists {
+				val += " (local)   "
+			} else {
+				val += " (upstream)"
+			}
+
+			if _, exists := bundles[set[bundle].Name]; exists {
+				val += " (included)"
+			}
+
+			fmt.Println(val)
+		}
+	case LocalList:
+		// Only print the top-level set
+		sorted := getBundleSetKeysSorted(bundles)
+		for _, bundle := range sorted {
+			val := fmt.Sprintf("%-*s", max, bundle)
+
+			if _, exists := mixBundles[bundle]; exists {
+				val += " (in mix)"
```

Comment:
Optional: Is it worth saying whether it masks an upstream bundle too? (at this point I'd say just hammer out and load all those lists hehe)

Reply:
So you can tell, when looking at local bundles, which are original and which are edited from upstream? I like that.

"Split up the list loading into the switch!", they say. "No, take some out and leave others in!", they say. "No, take them all back out!", they say. :wink: :stuck_out_tongue_winking_eye: 

In full seriousness, I think this extra level of detail is really useful for people understanding what's going on in their mix. I love all the ideas!